python/cudf/dataframe/categorical.py

```diff
@@ -258,26 +258,37 @@ def fillna(self, fill_value, inplace=False):
         """
         Fill null values with *fill_value*
         """
+
         result = self.copy()
 
-        if np.isscalar(fill_value):
-            if fill_value != self.default_na_value():
-                if (fill_value not in self.cat().categories):
-                    raise ValueError("fill value must be in categories")
-            fill_value = pd.Categorical(fill_value,
-                                        categories=self.cat().categories)
+        if not self.has_null_mask:
+            return result
+
+        fill_is_scalar = np.isscalar(fill_value)
+
+        if fill_is_scalar and fill_value == self.default_na_value():
+            out = cudautils.fillna(data=self.data.to_gpu_array(),
+                                   mask=self.mask.to_gpu_array(),
+                                   value=fill_value)
```

Comment:
I don't want us to regress to using this function. Have you identified what the expensive part in the below is that you're able to shortcircuit? Maybe we can just optimize that instead.

Reply:
@kkraus14 it looks like about half the time was spent in `_set_categories()`, which might be helped by a check like this:
```python
         if new_categories is self._categories:
            return self._parent.copy()
```
I couldn't see directly where the rest of the time went.

Reply:
@kkraus14 I copied the fillna implementation in `column.py`, but if there's a new/simpler method to replace the null values with -1, I'm happy to call it instead. The only optimization that matters is only touching the data buffer.

Reply:
As far as libcudf is concerned it only ever touches the data buffer, it has no concept of the categories in this sense. You should be able to pass a NumericalColumn of the same dtype as the category codes into the `replace_nulls` call and it should work as expected.

Reply:
@kkraus14 fixed